node/src/components/contract_runtime/rewards.rs

```diff
@@ -0,0 +1,380 @@
+                .height()
+                .saturating_sub(chainspec.core_config.signature_rewards_max_delay)
+        };
+        let range_to_fetch = cited_block_height_start..switch_block_height;
+
+        tracing::info!(
+            "[era {}] block range to fetch: {:?}",
+            era_id.value(),
+            range_to_fetch
+        );
+
+        collect_past_blocks_batched(effect_builder, range_to_fetch).await
+    };
+
+    //tracing::info!("blocks actually fetched: {cited_blocks:?}");
+    tracing::info!(
+        "[era {}] blocks fetched: {}/{}",
+        era_id.value(),
+        cited_blocks.iter().flatten().count(),
+        cited_blocks.len(),
+    );
+
+    let eras_info = {
+        let mut cited_blocks = cited_blocks.iter().flatten();
+        // We take the first block, then every switch block to get all the needed hashes to fetch
+        // needed eras:
+        let hashes_and_eras: Vec<_> = cited_blocks
+            .next()
+            .into_iter()
+            .chain(cited_blocks.filter(|&block| block.is_switch()))
+            .map(|block| {
+                (
+                    // The validator info for a switch block is the one from the next era:
+                    if block.is_switch() {
+                        block.era_id().successor()
+                    } else {
+                        block.era_id()
+                    },
+                    *block.hash(),
+                )
+            })
+            .collect();
+
+        tracing::info!(
+            "[era {}] will try and fetch {} eras",
+            era_id.value(),
+            hashes_and_eras.len()
+        );
+
+        ErasInfo::populate(
+            effect_builder,
+            chainspec.protocol_version(),
+            hashes_and_eras,
+        )
+        .await
+        .map_err(RewardsError::PopulateError)?
+    };
+
+    tracing::info!(
+        "[era {}] fetched eras: {:?}",
+        era_id.value(),
+        eras_info.list_of_eras_id()
+    );
+
+    let collection_proportion =
+        to_ratio_u512(chainspec.core_config.collection_rewards_proportion());
+    let contribution_proportion =
+        to_ratio_u512(chainspec.core_config.contribution_rewards_proportion());
+
+    // Reward for producing a block from this era:
+    let production_reward = to_ratio_u512(chainspec.core_config.production_rewards_proportion())
+        * eras_info.reward_pot(era_id)?;
+
+    // Collect all rewards as a ratio:
+    let full_reward_for_validators = {
+        let mut result = BTreeMap::new();
+
+        let blocks_from_current_era = cited_blocks.iter().filter_map(|maybe_block_with_metadata| {
+            maybe_block_with_metadata
+                .as_ref()
+                .filter(|block| block.era_id() == era_id)
+        });
+
+        for block in blocks_from_current_era {
+            // Reward for gathering all the finality signatures for this block:
+            let collection_reward = eras_info.weight_ratio(era_id, block.proposer())?
```

Comment:
This needs to be proportional to the weight of signatures reported in the block, also

Reply:
I did to the weight. We'll see if we need to change that.